### DISCUSSION TITLE
atomic_list_concat/2-3?

### DISCUSSION THREAD

**Discussion by PaulBrownMagic** [LogtalkDotOrg/logtalk3#102] [Ideas]
The `atomic_list_concat/2-3` is a rather popular predicate, but it's non-ISO so we can't depend on our backends to have it implemented.

How about we add them to an appropriate object? Maybe `atomic::list_concat/2-3` or `list(atomic)::concat/2-3`?

**Comment by pmoura** [comment]
There are several examples of what you suggest. E.g. The `termp` protocol declares the de facto standard `numbervars/3` predicate. The `integer` object declares the `between/3`, `plus/3`, and `succ/2` de facto standard predicates. Thus, we could add `atomic_list_concat/2-3` predicates to, say, the `atom` object. The objects implementing these predicates use the conditional compilation directives to use the backend built-in predicates when available. Static binding takes care of ensuring zero overhead. Typical usage would the then be:

```logtalk
:- uses(atom, [
    atomic_list_concat/2
]).
```

Another possible solution is to add them as a requirement for the adapters files. For those backends that don't (yet) provide them, we can then provide Prolog definitions.

But in any case, is preferable here to keep the predicate names and work with Prolog implementers in adding them if missing. So, the first step here is to write portable Prolog definitions for the predicates.

**Reply by PaulBrownMagic** [reply]
Well I found SWI has `atomic_list_concat/2-3`, but it doesn't concat with `[]`, which is atomic. ECLiPSe has `concat_atom/2` but for the arity 3 version it needs to go via a string. For GNU-Prolog I wrote something quick that does it via coercing numbers to atoms via `number_codes/2` and `atom_codes/2` and coercing `[]` & `{}` manually. (But that didn't work in ECLiPSe because they don't have `number_codes/2`, but it does work in `eclipselgt`, so good job on the adapter file!).

So it looks like I could write a portable version based on what I did for GNU testing; it works in `swilgt` and `eclipselgt` too.

**Reply by pmoura** [reply]
In SWI-Prolog, the empty list its a singleton type:

```text
?- atom([]).
false.

?- atomic([]).
true.
```

The ISO Prolog standard does require the empty list to be an atom. What's exactly the issue that you have found?

**Reply by PaulBrownMagic** [reply]
?- atomic([]).
    true.

    ?- atomic_concat(atomic, [], _).
    ERROR: Type error: `atom' expected, found `[]' (an empty_list)

But it's `atomic_concat/3`, not `atom_concat/3`.

**Reply by PaulBrownMagic** [reply]
Also:
    ?- atom_concat(a, 1, C).
    C = a1.
But 1 is not an atom. It seems like they're treated equivalently despite their names. This at least is documented in SWI.

**Reply by pmoura** [reply]
Ah, ok. I misunderstood as an issue in the `atomic_list_concat/2-3` predicates.

**Reply by PaulBrownMagic** [reply]
I've put a draft into a repo and invited you to it. I've no idea how it would be integrated into Logtalk. I'm rather proud of the tests because I finally got a good reason to use quick check!
